server/blob/blob-deduplicating/src/test/scala/org/apache/james/server/blob/deduplication/GCPropertiesTest.scala

```diff
@@ -0,0 +1,215 @@
+
+  def testParametersGen(eventsGen: Gen[Seq[Event]]): Gen[TestParameters] = for {
+    events <- eventsGen
+    allGenerations = Generation.range(Generation.first, Events.getLastGeneration(events))
+    generationsToCollect <- Gen.someOf(allGenerations)
+  } yield TestParameters(events, generationsToCollect.sorted.toSeq)
+
+  def onePassTestParametersGen(eventsGen: Gen[Seq[Event]]): Gen[OnePassGCTestParameters] = for {
+    events <- eventsGen
+    generation <- Gen.oneOf(if(events.isEmpty) Seq(Generation.first) else events.map(_.generation).toSet)
+  } yield OnePassGCTestParameters(events, generation)
+}
+
+object GCPropertiesTest extends Properties("GC") {
+  val maxNbEvents = 100
+  val hashesPerEventsRatio = 0.2f
+
+  val arbEvents: Arbitrary[Seq[Event]] = Arbitrary(Gen.choose(0, maxNbEvents).flatMap(Generators.eventsGen(_, hashesPerEventsRatio)))
+  implicit val arbTestParameters: Arbitrary[Generators.TestParameters] = Arbitrary(Generators.testParametersGen(arbEvents.arbitrary))
+  implicit val arbTestParameter: Arbitrary[Generators.OnePassGCTestParameters] = Arbitrary(Generators.onePassTestParametersGen(arbEvents.arbitrary))
+  import org.scalacheck.Shrink._
+
+  override def overrideParameters(p: Parameters) =
+    p.withMinSuccessfulTests(1000)
+
+  def createSaneTestParameters(events: Seq[Event], generations: Seq[Generation]): TestParameters = {
+    val allGenerations = Generation.range(Generation.first, Events.getLastGeneration(events))
+    TestParameters(events, generations.filter(allGenerations.contains(_)))
+  }
+  def createSaneOnePassGCTestParameters(events: Seq[Event], generation: Generation): OnePassGCTestParameters = {
+    OnePassGCTestParameters(events, generation)
+  }
+
+  implicit val shrinkTestParameters: Shrink[Generators.TestParameters] = Shrink {
+    params: Generators.TestParameters =>
+      shrink(params.events).flatMap(events => shrink(params.generationsToCollect).map(generations => createSaneTestParameters(events, generations)))
+  }
+
+  implicit val shrinkOnePassGCTestParameters: Shrink[Generators.OnePassGCTestParameters] = Shrink {
+    params: Generators.OnePassGCTestParameters =>
+      shrink(params.events).map(events => createSaneOnePassGCTestParameters(events, params.generationToCollect))
+  }
+
+  property("2.1. GC should not delete data being referenced by a pending process or still referenced") = forAll {
+    testParameters: Generators.TestParameters => {
+
+      val (stillReferencedBlobIds, _) = partitionBlobs(testParameters.events)
+      testParameters.generationsToCollect.foldLeft(true)((acc, e) => {
+        val plannedDeletions = GC.plan(Interpreter(testParameters.events).stabilize(), Iteration.initial, e).blobsToDelete.map(_._2)
+        acc && stillReferencedBlobIds.intersect(plannedDeletions).isEmpty
+      })
+    }
+  }
+
+  property("3.2. less than 10% of unreferenced data of a significant dataset should persist") = forAll {
```

Comment:
Question: what's the actual mean value for 1000 runs (just curious)

Reply:
0 for the moment, we use accurate data structures.